File: packages/runar-compiler/src/__tests__/unknown-anf-kind.test.ts
/**
 * Regression test for F-003: every ANF-kind dispatch in the TS compiler
 * must throw UnknownANFKindError when it encounters a kind it doesn't
 * recognize, instead of silently returning an empty / no-op result.
 *
 * Each test drives one dispatch site with a synthetic ANFValue whose
 * `kind` does not appear in the ANFValue union, then asserts the
 * resulting throw is the typed error and carries the synthetic kind name.
 *
 * If a new ANFValue variant is added in the future, the dispatch sites
 * below must be updated; this test guards against silently shipping an
 * unhandled variant.
 */

import { describe, it, expect } from 'vitest';
import { UnknownANFKindError } from 'runar-ir-schema';
import type {
  ANFProgram,
  ANFMethod,
  ANFBinding,
  ANFValue,
} from '../ir/index.js';
import { foldConstants, eliminateDeadBindings } from '../optimizer/constant-fold.js';
import { lowerToStack } from '../passes/05-stack-lower.js';
import { remapValueRefs } from '../passes/04-anf-lower.js';

const SYNTHETIC_KIND = 'synthetic_test_kind_for_regression_only';

/**
 * Build an ANFValue that the type system thinks is fine but whose runtime
 * `kind` discriminator is not in the schema. The cast is intentional —
 * the whole point is to simulate a developer adding a new ANF variant
 * without wiring it through every dispatch site.
 */
function syntheticValue(): ANFValue {
  return { kind: SYNTHETIC_KIND } as unknown as ANFValue;
}

function makeProgram(body: ANFBinding[]): ANFProgram {
  const method: ANFMethod = {
    name: 'm',
    params: [],
    body,
    isPublic: true,
  };
  return {
    contractName: 'Test',
    properties: [],
    methods: [method],
  };
}

describe('UnknownANFKindError — dispatch-site regression guard', () => {
  it('throws from constant-fold.foldValue', () => {
    const program = makeProgram([{ name: 't0', value: syntheticValue() }]);

    let caught: unknown;
    try {
      foldConstants(program);
    } catch (e) {
      caught = e;
    }

    expect(caught).toBeInstanceOf(UnknownANFKindError);
    expect((caught as UnknownANFKindError).kind).toBe(SYNTHETIC_KIND);
    expect((caught as UnknownANFKindError).location).toBe('constant-fold.foldValue');
  });

  it('throws from constant-fold.collectRefsFromValue (via DCE)', () => {
    // hasSideEffect is consulted first in filterLiveBindings; we want to
    // hit collectRefsFromValue, which runs over the body before the
    // side-effect filter. Putting the synthetic binding inside an `if`
    // forces the dead-binding pass to walk into nested values via
    // collectRefsFromValue's recursive case before hasSideEffect sees it.
    const program = makeProgram([
      { name: 't0', value: { kind: 'load_const', value: true } },
      {
        name: 't1',
        value: {
          kind: 'if',
          cond: 't0',
          then: [{ name: 'tn', value: syntheticValue() }],
          else: [],
        },
      },
    ]);

    let caught: unknown;
    try {
      eliminateDeadBindings(program);
    } catch (e) {
      caught = e;
    }

    expect(caught).toBeInstanceOf(UnknownANFKindError);
    expect((caught as UnknownANFKindError).kind).toBe(SYNTHETIC_KIND);
    // The first dispatch to fail is collectRefsFromValue inside the if-walk.
    expect((caught as UnknownANFKindError).location).toBe(
      'constant-fold.collectRefsFromValue',
    );
  });

  it('throws from constant-fold.hasSideEffect (via DCE on flat body)', () => {
    // With a flat body, collectAllRefs walks only top-level bindings via
    // the switch (load_const path is hit first, with no nested values).
    // The synthetic binding has no nested refs so collectRefsFromValue
    // throws there too — but for the side-effect-only path we craft a
    // body where the synthetic value is the only binding, so the first
    // dispatch hit is collectRefsFromValue. To actually reach
    // hasSideEffect we'd have to skip ref-collection; instead we assert
    // that *something* in the DCE pipeline rejects the synthetic kind.
    const program = makeProgram([{ name: 't0', value: syntheticValue() }]);

    let caught: unknown;
    try {
      eliminateDeadBindings(program);
    } catch (e) {
      caught = e;
    }

    expect(caught).toBeInstanceOf(UnknownANFKindError);
    expect((caught as UnknownANFKindError).kind).toBe(SYNTHETIC_KIND);
    // Either dispatch site is acceptable — both must reject the kind.
    expect([
      'constant-fold.collectRefsFromValue',
      'constant-fold.hasSideEffect',
    ]).toContain((caught as UnknownANFKindError).location);
  });

  it('throws from stack-lower (collectRefs or lowerBinding)', () => {
    const program = makeProgram([{ name: 't0', value: syntheticValue() }]);

    let caught: unknown;
    try {
      lowerToStack(program);
    } catch (e) {
      caught = e;
    }

    expect(caught).toBeInstanceOf(UnknownANFKindError);
    expect((caught as UnknownANFKindError).kind).toBe(SYNTHETIC_KIND);
    // collectRefs runs first (computeLastUses) — that's where we expect
    // the throw — but lowerBinding is the fallback. Both are acceptable.
    expect([
      'stack-lower.collectRefs',
      'stack-lower.lowerBinding',
    ]).toContain((caught as UnknownANFKindError).location);
  });

  it('throws from anf-lower.remapValueRefs', () => {
    let caught: unknown;
    try {
      remapValueRefs(syntheticValue(), {});
    } catch (e) {
      caught = e;
    }

    expect(caught).toBeInstanceOf(UnknownANFKindError);
    expect((caught as UnknownANFKindError).kind).toBe(SYNTHETIC_KIND);
    expect((caught as UnknownANFKindError).location).toBe('anf-lower.remapValueRefs');
  });

  it('carries an actionable message that references the developer recipe', () => {
    const err = new UnknownANFKindError(SYNTHETIC_KIND, 'unit-test.location');
    expect(err.message).toContain(SYNTHETIC_KIND);
    expect(err.message).toContain('unit-test.location');
    expect(err.message).toContain('Adding a New ANF Value Kind');
    expect(err.name).toBe('UnknownANFKindError');
  });
});

File: compilers/go/codegen/unknown_anf_kind_test.go
// Package codegen — regression test for F-003: every ANF-kind dispatch
// in the Go stack-lowering codegen must reject an unknown kind with a
// typed *ir.UnknownANFKindError, instead of silently returning empty
// refs / emitting no stack ops.
//
// Each test drives one dispatch site with a synthetic ANFValue whose
// `Kind` does not appear in the ANFValue schema, then asserts the
// resulting panic carries the typed error with the synthetic kind name.
//
// If a new ANFValue variant is added in the future, the dispatch sites
// below must be updated; this test guards against silently shipping an
// unhandled variant.
package codegen

import (
	"errors"
	"strings"
	"testing"

	"github.com/icellan/runar/compilers/go/ir"
)

const syntheticKind = "synthetic_test_kind_for_regression_only"

func syntheticANFValue() *ir.ANFValue {
	return &ir.ANFValue{Kind: syntheticKind}
}

// recoverUnknownKind runs fn and recovers a panic. It returns the typed
// error if the panic value was *ir.UnknownANFKindError, otherwise nil.
func recoverUnknownKind(t *testing.T, fn func()) *ir.UnknownANFKindError {
	t.Helper()
	var caught *ir.UnknownANFKindError
	func() {
		defer func() {
			if r := recover(); r != nil {
				if e, ok := r.(*ir.UnknownANFKindError); ok {
					caught = e
					return
				}
				t.Fatalf("expected *ir.UnknownANFKindError panic, got %T: %v", r, r)
			}
		}()
		fn()
	}()
	return caught
}

func TestUnknownANFKind_Stack_CollectRefs(t *testing.T) {
	err := recoverUnknownKind(t, func() {
		_ = collectRefs(syntheticANFValue())
	})
	if err == nil {
		t.Fatal("expected panic with *ir.UnknownANFKindError, got none")
	}
	if err.Kind != syntheticKind {
		t.Errorf("expected kind %q, got %q", syntheticKind, err.Kind)
	}
	if err.Location != "stack.collectRefs" {
		t.Errorf("expected location stack.collectRefs, got %q", err.Location)
	}
}

func TestUnknownANFKind_Stack_LowerBinding_ViaLowerToStack(t *testing.T) {
	// LowerToStack wraps panics in a returned error via fmt.Errorf("...: %v", r),
	// so the typed error surfaces through its Error() method on the wrapped value.
	program := &ir.ANFProgram{
		ContractName: "T",
		Methods: []ir.ANFMethod{
			{
				Name:     "m",
				Params:   nil,
				IsPublic: true,
				Body: []ir.ANFBinding{
					{Name: "t0", Value: ir.ANFValue{Kind: syntheticKind}},
				},
			},
		},
	}
	_, err := LowerToStack(program)
	if err == nil {
		t.Fatal("expected LowerToStack to return an error for an unknown kind")
	}
	if !strings.Contains(err.Error(), syntheticKind) {
		t.Errorf("expected error to contain synthetic kind %q, got %q", syntheticKind, err.Error())
	}
	// Either collectRefs (computeLastUses runs first) or lowerBinding can fire.
	if !strings.Contains(err.Error(), "stack.collectRefs") &&
		!strings.Contains(err.Error(), "stack.lowerBinding") {
		t.Errorf("expected stack.collectRefs or stack.lowerBinding in error, got %q", err.Error())
	}
}

func TestUnknownANFKindError_MessageMentionsRecipe(t *testing.T) {
	e := &ir.UnknownANFKindError{Kind: syntheticKind, Location: "unit-test.location"}
	msg := e.Error()
	if !strings.Contains(msg, syntheticKind) {
		t.Errorf("expected message to contain kind, got %q", msg)
	}
	if !strings.Contains(msg, "unit-test.location") {
		t.Errorf("expected message to contain location, got %q", msg)
	}
	if !strings.Contains(msg, "Adding a New ANF Value Kind") {
		t.Errorf("expected message to point at CLAUDE.md recipe, got %q", msg)
	}
	// The type should be usable as a standard error.
	var target *ir.UnknownANFKindError
	if !errors.As(error(e), &target) {
		t.Errorf("errors.As did not match *ir.UnknownANFKindError")
	}
}
